File: Samples/AddPageToTiff/C#/AddPageToTiff.cs
using System;
using System.Diagnostics;
using System.Drawing;
using System.Drawing.Imaging;
using System.IO;

using BitMiracle.LibTiff.Classic;

namespace BitMiracle.LibTiff.Samples
{
    public static class AddPageToTiff
    {
        public static void Main()
        {
            File.Copy(@"Sample Data\16bit.tif", @"Sample Data\ToBeAppended.tif", true);

            using (Tiff image = Tiff.Open(@"Sample Data\ToBeAppended.tif", "a"))
            {
                int newPageNumber = image.NumberOfDirectories() + 1;
                const int width = 100;
                const int height = 100;

                image.SetField(TiffTag.IMAGEWIDTH, width);
                image.SetField(TiffTag.IMAGELENGTH, height);
                image.SetField(TiffTag.BITSPERSAMPLE, 8);
                image.SetField(TiffTag.SAMPLESPERPIXEL, 3);
                image.SetField(TiffTag.ROWSPERSTRIP, height);

                image.SetField(TiffTag.COMPRESSION, Compression.LZW);
                image.SetField(TiffTag.PHOTOMETRIC, Photometric.RGB);
                image.SetField(TiffTag.FILLORDER, FillOrder.MSB2LSB);
                image.SetField(TiffTag.PLANARCONFIG, PlanarConfig.CONTIG);

                byte[] buffer = null;
                using (Bitmap bmp = new Bitmap(width, height, PixelFormat.Format24bppRgb))
                {
                    using (Graphics g = Graphics.FromImage(bmp))
                    {
                        g.FillRectangle(Brushes.White, g.VisibleClipBounds);
                        string s = newPageNumber.ToString();
                        Font f = SystemFonts.DefaultFont;

                        SizeF size = g.MeasureString(s, f);
                        PointF loc = new PointF(Math.Max((bmp.Width - size.Width) / 2, 0), Math.Max((bmp.Height - size.Height) / 2, 0));
                        g.DrawString(s, f, Brushes.Black, loc);

                        buffer = getImageRasterBytes(bmp, PixelFormat.Format24bppRgb);
                    }
                }

                int stride = buffer.Length / height;
                convertRGBSamples(buffer, width, height);

                for (int i = 0, offset = 0; i < height; i++)
                {
                    image.WriteScanline(buffer, offset, i, 0);
                    offset += stride;
                }
            }

            Process.Start(@"Sample Data\ToBeAppended.tif");
        }

        private static byte[] getImageRasterBytes(Bitmap bmp, PixelFormat format)
        {
            Rectangle rect = new Rectangle(0, 0, bmp.Width, bmp.Height);
            byte[] bits = null;

            try
            {
                // Lock the managed memory
                BitmapData bmpdata = bmp.LockBits(rect, ImageLockMode.ReadWrite, format);

                // Declare an array to hold the bytes of the bitmap.
                bits = new byte[bmpdata.Stride * bmpdata.Height];

                // Copy the values into the array.
                System.Runtime.InteropServices.Marshal.Copy(bmpdata.Scan0, bits, 0, bits.Length);

                // Release managed memory
                bmp.UnlockBits(bmpdata);
            }
            catch
            {
                return null;
            }

            return bits;
        }

        /// <summary>
        /// Converts BGR samples into RGB samples
        /// </summary>
        private static void convertRGBSamples(byte[] data, int width, int height)
        {
            int stride = data.Length / height;
            const int samplesPerPixel = 3;

            for (int y = 0; y < height; y++)
            {
                int offset = stride * y;
                int strideEnd = offset + width * samplesPerPixel;

                for (int i = offset; i < strideEnd; i += samplesPerPixel)
                {
                    byte temp = data[i + 2];
                    data[i + 2] = data[i];
                    data[i] = temp;
                }
            }
        }
    }
}

File: Samples/AddPageToTiff/VB.NET/AddPageToTiff.vb
Imports System.Diagnostics
Imports System.Drawing
Imports System.Drawing.Imaging
Imports System.IO

Imports BitMiracle.LibTiff.Classic

Namespace BitMiracle.LibTiff.Samples
    Public NotInheritable Class AddPageToTiff
        Private Sub New()
        End Sub
        Public Shared Sub Main()
            File.Copy("Sample Data\16bit.tif", "Sample Data\ToBeAppended.tif", True)

            Using image As Tiff = Tiff.Open("Sample Data\ToBeAppended.tif", "a")
                Dim newPageNumber As Integer = image.NumberOfDirectories() + 1
                Const width As Integer = 100
                Const height As Integer = 100

                image.SetField(TiffTag.IMAGEWIDTH, width)
                image.SetField(TiffTag.IMAGELENGTH, height)
                image.SetField(TiffTag.BITSPERSAMPLE, 8)
                image.SetField(TiffTag.SAMPLESPERPIXEL, 3)
                image.SetField(TiffTag.ROWSPERSTRIP, height)

                image.SetField(TiffTag.COMPRESSION, Compression.LZW)
                image.SetField(TiffTag.PHOTOMETRIC, Photometric.RGB)
                image.SetField(TiffTag.FILLORDER, FillOrder.MSB2LSB)
                image.SetField(TiffTag.PLANARCONFIG, PlanarConfig.CONTIG)

                Dim buffer As Byte() = Nothing
                Using bmp As New Bitmap(width, height, PixelFormat.Format24bppRgb)
                    Using g As Graphics = Graphics.FromImage(bmp)
                        g.FillRectangle(Brushes.White, g.VisibleClipBounds)
                        Dim s As String = newPageNumber.ToString()
                        Dim f As Font = SystemFonts.DefaultFont

                        Dim size As SizeF = g.MeasureString(s, f)
                        Dim loc As New PointF(Math.Max((bmp.Width - size.Width) / 2, 0), Math.Max((bmp.Height - size.Height) / 2, 0))
                        g.DrawString(s, f, Brushes.Black, loc)

                        buffer = getImageRasterBytes(bmp, PixelFormat.Format24bppRgb)
                    End Using
                End Using

                Dim stride As Integer = buffer.Length \ height
                convertRGBSamples(buffer, width, height)

                Dim i As Integer = 0, offset As Integer = 0
                While i < height
                    image.WriteScanline(buffer, offset, i, 0)
                    offset += stride
                    i += 1
                End While
            End Using

            Process.Start("Sample Data\ToBeAppended.tif")
        End Sub

        Private Shared Function getImageRasterBytes(ByVal bmp As Bitmap, ByVal format As PixelFormat) As Byte()
            Dim rect As New Rectangle(0, 0, bmp.Width, bmp.Height)
            Dim bits As Byte() = Nothing

            Try
                ' Lock the managed memory
                Dim bmpdata As BitmapData = bmp.LockBits(rect, ImageLockMode.ReadWrite, format)

                ' Declare an array to hold the bytes of the bitmap.
                bits = New Byte(bmpdata.Stride * bmpdata.Height - 1) {}

                ' Copy the values into the array.
                System.Runtime.InteropServices.Marshal.Copy(bmpdata.Scan0, bits, 0, bits.Length)

                ' Release managed memory
                bmp.UnlockBits(bmpdata)
            Catch
                Return Nothing
            End Try

            Return bits
        End Function

        ''' <summary>
        ''' Converts BGR samples into RGB samples
        ''' </summary>
        Private Shared Sub convertRGBSamples(ByVal data As Byte(), ByVal width As Integer, ByVal height As Integer)
            Dim stride As Integer = data.Length \ height
            Const samplesPerPixel As Integer = 3

            For y As Integer = 0 To height - 1
                Dim offset As Integer = stride * y
                Dim strideEnd As Integer = offset + width * samplesPerPixel

                Dim i As Integer = offset
                While i < strideEnd
                    Dim temp As Byte = data(i + 2)
                    data(i + 2) = data(i)
                    data(i) = temp
                    i += samplesPerPixel
                End While
            Next
        End Sub
    End Class
End Namespace
